### PR TITLE
Multi Pin Bumps across PT/AO/tune/ET

### PR DESCRIPTION
**Pull request by Jack-Khuu**
Accounts for: 
- PyTorch changing `weight_only` default from `False` to `True` https://github.com/pytorch/torchchat/issues/1356
- ~~Moving from `export` to `export_for_training` https://github.com/pytorch/torchchat/pull/1319~~
  - Fixed in https://github.com/pytorch/torchchat/pull/1419
- ~~Should also fix cuDNN error: https://github.com/huggingface/diffusers/issues/9704~~ 
  - Resolution indirect from this PR 
- _convert_weight_to_int4pack API change from https://github.com/pytorch/pytorch/pull/139611, requiring AO PinBump for https://github.com/pytorch/ao/pull/1278
- Change in CUDA support in PT wheels https://github.com/pytorch/pytorch/issues/140885
- ~~Lock on llama.cpp [SHA](https://github.com/ggerganov/llama.cpp/commit/64ed2091b24b2f9747148fdf49a34ed5938762c3) to avoid active changes (e.g. deprecating make https://github.com/ggerganov/llama.cpp/pull/10514)~~
  - Fixed in https://github.com/pytorch/torchchat/pull/1400

- PT buildwheel migration https://github.com/pytorch/pytorch/issues/123649
- Missing OMP in nightly regression on aarch64 https://github.com/pytorch/pytorch/issues/142266

---
**Update:** 
Merging while accepting the errors in
* `*-gpu-aoti`, `aot-cuda`: Due to byte alignment related to https://github.com/pytorch/pytorch/pull/140664. Fix in progress and will be picked up in separate pin bump https://github.com/pytorch/pytorch/pull/143236
* `compile-gguf`: Working on updating the adoption in torchchat https://github.com/pytorch/torchchat/pull/1404

### CODE REVIEW

**Comment by swolchok** [comment]
initial debugging shows the test-cpu-aoti segfault is within aoti_torch_cpu_cat, which is automatically generated by https://github.com/pytorch/pytorch/blob/7e86a7c0155295539996e0cf422883571126073e/torchgen/gen_aoti_c_shim.py . digging up the generated source now.

File: torchchat/distributed/checkpoint.py [hunk 1]
```diff
@@ -96,6 +96,7 @@ def _load_checkpoints_from_storage(
         checkpoint_path,
         map_location=builder_args.device,
         mmap=True,
+        weight_only=False,
```
Comment (malfet) [inline torchchat/distributed/checkpoint.py:99]: Why does it needs false? All LLMs should be loadable with weights_only, shouldn't they? (Also, there are no such option as `weight_only` (or so I hope :P ))
```suggestion
        weights_only=True,
```

File: torchchat/export.py [hunk 2]
```diff
@@ -238,7 +238,7 @@ def _to_core_aten(
             raise ValueError(
                 f"Expected passed in model to be an instance of fx.GraphModule, got {type(model)}"
             )
-        core_aten_ep = export(model, example_inputs, dynamic_shapes=dynamic_shapes)
+        core_aten_ep = export_for_training(model, example_inputs, dynamic_shapes=dynamic_shapes)
```
Comment (malfet) [inline torchchat/export.py:241]: Not sure what we are doing here, but shouldn't TorchChat be exporting for inference?

**Review by malfet — APPROVED** [review]

**Comment by swolchok** [comment]
> digging up the generated source now.

generated source looks OK. here's what doesn't look OK in the generated inductor .cpp file:

```
    AtenTensorHandle buf0_handle;
    AOTI_TORCH_ERROR_CODE_CHECK(aoti_torch_empty_strided(2, int_array_12, int_array_13, cached_torch_dtype_uint8, cached_torch_device_type_cpu, this->device_idx_, &buf0_handle));
    RAIIAtenTensorHandle buf0(buf0_handle);
    AtenTensorHandle buf1_handle;
    AOTI_TORCH_ERROR_CODE_CHECK(aoti_torch_empty_strided(2, int_array_12, int_array_13, cached_torch_dtype_uint8, cached_torch_device_type_cpu, this->device_idx_, &buf1_handle));
    RAIIAtenTensorHandle buf1(buf1_handle);
    cpp_fused_div_remainder_0((const uint8_t*)(self___model_tok_embeddings__buffers__weight.data_ptr()), (uint8_t*)(buf0.data_ptr()), (uint8_t*)(buf1.data_ptr()));
    // Topologically Sorted Source Nodes: [weight_unpacked], Original ATen: [aten.stack]
    static constexpr int64_t int_array_0[] = {32000LL, 144LL, 1LL};
    static constexpr int64_t int_array_1[] = {144LL, 1LL, 0LL};
    auto tmp_tensor_handle_0 = reinterpret_tensor_wrapper(buf0, 3, int_array_0, int_array_1, 0LL);
    auto tmp_tensor_handle_1 = reinterpret_tensor_wrapper(buf1, 3, int_array_0, int_array_1, 0LL);
    const AtenTensorHandle var_array_0[] = {wrap_with_raii_handle_if_needed(tmp_tensor_handle_0), wrap_with_raii_handle_if_needed(tmp_tensor_handle_1)};
    AtenTensorHandle buf3_handle;
    AOTI_TORCH_ERROR_CODE_CHECK(aoti_torch_cpu_cat(var_array_0, 2, -1LL, &buf3_handle));
```

The problem seems to be `   const AtenTensorHandle var_array_0[] = {wrap_with_raii_handle_if_needed(tmp_tensor_handle_0), wrap_with_raii_handle_if_needed(tmp_tensor_handle_1)};` -- this is creating RAIIATenTensorHandles, whose `operator ATenTensorHandle` is immediately called, and then they're destroyed (which decrements the refcount), so the net effect is (I think) to create dangling ATenTensorHandles.

**Comment by swolchok** [comment]
@desertfire any change the above is a quick fix for you?

**Comment by swolchok** [comment]
actually we might just need https://github.com/pytorch/pytorch/pull/139411

File: torchchat/export.py (hunk 2 above)
Comment (Jack-Khuu) [reply torchchat/export.py:241]: This was picked up from @tugsbayasgalan's PR migrating away from export(), but export_for_inference does sound  more in line with what we want

@tugsbayasgalan Can you share info on the new APIs?

File: torchchat/distributed/checkpoint.py (hunk 1 above)
Comment (Jack-Khuu) [reply torchchat/distributed/checkpoint.py:99]: Good catch on the typo; 

As for setting it to False: I'd rather keep it behavior consistent in a pin bump PR; we can flip in a separate PR

File: torchchat/export.py (hunk 2 above)
Comment (tugsbayasgalan) [reply torchchat/export.py:241]: Yep the intended use for inference IR is that user will export to a training IR and call run_decompositions() to lower to inference IR. In this flow, after core_aten_ep, there is to_edge call which lowers to inference. Export team is moving the IR to non-functional training IR so export_for_training will exist as an alias to official export. After we actually migrate official export, we will replace this call with export.

**Comment by swolchok** [comment]
no torchvision nightly again today. I'm guessing we could probably use torchvision from yesterday with torch from today?

**Comment by Jack-Khuu** [comment]
I had issues with Vision nightlies requiring the corresponding PT nightly few weeks back, I'll give it another go

**Comment by swolchok** [comment]
_convert_weight_to_int4pack breakage appears to be from https://github.com/pytorch/pytorch/pull/139611; I guess it's now called _convert_weight_to_int4pack_for_cpu .

**Comment by Jack-Khuu** [comment]
AO has a fix so we'll need a bump there too: https://github.com/pytorch/ao/pull/1278

**Comment by Jack-Khuu** [comment]
https://github.com/pytorch/pytorch/pull/139411 Also got reverted on pt/pt so that's fun

**Comment by Jack-Khuu** [comment]
Need to bump everything cuda related: https://github.com/pytorch/pytorch/issues/140885

**Comment by swolchok** [comment]
> Best me to it; luckily AO has a fix so we'll need a bump there too: [pytorch/ao#1278](https://github.com/pytorch/ao/pull/1278)

also need to manually edit torchchat/utils/gguf_loader.py

**Comment by Jack-Khuu** [comment]
Bumped the AO/pt pin, will update the gguf_loader next

Looks like we need https://github.com/pytorch/pytorch/pull/141063

> ImportError: /home/jackkhuu/.conda/envs/test_D/lib/python3.10/site-packages/torch/lib/../../nvidia/cusparse/lib/libcusparse.so.12: undefined symbol: __nvJitLinkComplete_12_4, version libnvJitLink.so.12

**Comment by Jack-Khuu** [comment]
Need to bump tune version to pick up changes from AO BC changes (https://github.com/pytorch/ao/pull/1049)

**Comment by Jack-Khuu** [comment]
Potential AO import bug? https://github.com/pytorch/ao/issues/1365

**Comment by Jack-Khuu** [comment]
https://github.com/pytorch/vision/issues/8778

**Comment by Jack-Khuu** [comment]
> torchvision 0.20.0.dev20241205 depends on torch==2.6.0.dev20241204

Smells like a bug on the vision side, but testing mismatch version

**Comment by Jack-Khuu** [comment]
Talking to torchvision team about mismatch torch dep

**Comment by Jack-Khuu** [comment]
Missing omp.h for aarch is a regression on pt/pt, we're looking into it and will pinbump to a fix when found

File: .github/workflows/more-tests.yml [hunk 3]
```diff
@@ -9,23 +9,17 @@ on:
 
 jobs:
   test-cuda:
-    uses: pytorch/test-infra/.github/workflows/linux_job.yml@main
+    uses: pytorch/test-infra/.github/workflows/linux_job_v2.yml@main
```
Comment (larryliu0820) [inline .github/workflows/more-tests.yml:12]: What's the difference between these two?

Comment (Jack-Khuu) [reply .github/workflows/more-tests.yml:12]: Former is deprecated: https://github.com/pytorch/pytorch/issues/123649

**Comment by Jack-Khuu** [comment]
Note that the changes to gguf_loader are suspicious at best and an attempt to recover previous behavior prior to https://github.com/pytorch/pytorch/pull/139611

**Comment by Jack-Khuu** [comment]
> Missing omp.h for aarch is a regression on pt/pt, we're looking into it and will pinbump to a fix when found
> 
> [pytorch/pytorch#142266](https://github.com/pytorch/pytorch/issues/142266)

Hmmm looks like the wheels aren't fixed with the changes...

**Pull request merged**